File: intermediate-package/writing-integ-tests-for-spring-web-apps/rest-api/spring/src/main/java/com/testwithspring/intermediate/web/TaskErrorHandler.java
package com.testwithspring.intermediate.web;

import com.testwithspring.intermediate.common.NotFoundException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;

import java.util.List;

@ControllerAdvice
public class TaskErrorHandler {

    private static final Logger LOGGER = LoggerFactory.getLogger(TaskErrorHandler.class);

    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ResponseBody
    public ValidationErrorDTO processValidationError(MethodArgumentNotValidException ex) {
        BindingResult result = ex.getBindingResult();
        List<FieldError> fieldErrors = result.getFieldErrors();
        return processFieldErrors(fieldErrors);
    }

    private ValidationErrorDTO processFieldErrors(List<FieldError> fieldErrors) {
        ValidationErrorDTO dto = new ValidationErrorDTO();

        for (FieldError fieldError: fieldErrors) {
            String localizedErrorMessage = resolveErrorCode(fieldError);
            dto.addFieldError(fieldError.getField(), localizedErrorMessage);
        }

        return dto;
    }

    private String resolveErrorCode(FieldError fieldError) {
        String[] fieldErrorCodes = fieldError.getCodes();
        return fieldErrorCodes[fieldErrorCodes.length - 1];
    }

    @ExceptionHandler(NotFoundException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public void returnHttpStatusCodeNotFound() {
        LOGGER.error("Task was found not. Returning HTTP status code 404");
    }
}

File: master-package/bonus-webinars-2019/unit-tests-rest-api/src/main/kotlin/com/testwithspring/master/web/TaskErrorHandler.kt
package com.testwithspring.master.web

import com.testwithspring.master.common.NotFoundException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.ResponseStatus


/**
 * Handles the exceptions thrown by our web application.
 */
@ControllerAdvice
open class TaskErrorHandler {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(TaskErrorHandler::class.java)
    }

    @ExceptionHandler(NotFoundException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    open fun handleNotFoundError(ex: NotFoundException) {
        LOGGER.error("Handling exception with message: {}", ex.message)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ResponseBody
    open fun handleValidationError(ex: MethodArgumentNotValidException): ValidationErrorDTO {
        val result = ex.bindingResult
        val fieldErrors = result.fieldErrors
        return processFieldErrors(fieldErrors)
    }

    private fun processFieldErrors(fieldErrors: List<FieldError>): ValidationErrorDTO {
        val dto = ValidationErrorDTO()
        for (fieldError in fieldErrors) {
            val errorCode = resolveErrorCode(fieldError)
            dto.addFieldError(fieldError.field, errorCode)
        }
        return dto
    }

    private fun resolveErrorCode(fieldError: FieldError): String {
        val fieldErrorCodes = fieldError.codes
        return fieldErrorCodes?.get(fieldErrorCodes.size - 1) ?: "UnknownError"
    }

}